GitHub Actions builds and automated tags/releases

redtide opened the pull request:
In this PR GitHub Actions CI will produce different artifacts depending when in release or simple commits and pull requests.
- On normal commits or pull requests only portable builds are available as artifacts, using dynamic linking.
- On release lhelper is used to build static versions of Freetype, PCRE2 and SDL2 libraries.
  Portable builds are available as in normal commits, plus a Linux AppImage, Windows InnoSetup 32 (see issue below) and 64 bit packages and a source code tarball including subprojects for distro packagers will be produced.

In both cases a DMG image is made for macOS, using [appdmg](https://github.com/LinusU/node-appdmg) for better looking with background image and symlink to `Applications` for application bundle copy

Linux matrix builds still testing using gcc and clang, gcc only is used to produce the packages/artifacts.

Fixes #345.

TODO:
- Probably need to split in other PRs for semplicity.
- Releases deployment is temporarily disabled:
  - macOS: missing notarization data

francesco-st commented:
How to install lhelper:

```sh
wget -q -O - https://raw.githubusercontent.com/franko/lhelper/master/install-github | bash -
```

Then create an environment to compile lite-xl:

```sh
lhelper create lite-xl
# once created the environment is automatically activated
```

To create an environment using another compiler or provide some compiler flags:

```sh
CC=clang CXX=clang++ CFLAGS=<some flags> CXXFLAGS=<some flags> lhelper create lite-xl
# once created the environment is automatically activated
````

Then install libraries needed for lite-xl:

```sh
lhelper install freetype2
lhelper install sdl2 2.0.14-wait-event-timeout-1
lhelper install pcre2
```

at this point lite-xl can be compiled using Meson following the standard procedure. I recommend using the script build-package.sh but this is not mandatory.

As far as I know the lhelper operations above can run unattended in the server but I understand errors may appear or we may need some adaptations. I may do some modifications in lhelper if needed for CI integration.

## Not using lhelper

If we don't use lhelper to compile the libraries we need to do manually. We will never include these libraries as subprojects.

For example for SDL2 the steps:

- download the source
- run the configure script with the command (for windows)
```sh
configure --enable-video --enable-events --enable-timers --enable-assembly --enable-sdl2-config --disable-threads --disable-atomic --disable-dbus --disable-ibus --disable-fcitx --disable-directx --disable-render-d3d --disable-video-vulkan --disable-video-directfb --disable-video-dummy --disable-video-opengles --disable-video-opengles1 --disable-video-opengles2 --disable-video-vivante --disable-video-opengl --disable-loadso --disable-render --disable-audio --disable-joystick --disable-haptic --disable-hidapi --disable-sensor --disable-power --disable-filesystem --disable-file --disable-cpuinfo --disable-ime --disable-shared
```
- compile and install SDL2

Rinse and repeat for the other libraries. Please note that lhelper is just doing this job. Note also that of each OS we need to pass different options.

Another options would be to use:

```sh
configure  --disable-shared
```
but in this case all the options will be activated by default.

I propose instead that we fix the problem we may have to use lhelper in the CI script.

redtide commented:
Thanks, I was just about to write that I followed again the [procedure][1] you told me and this time I had no interactive issues, so maybe I was doing something wrong.
I think we can use lhelper and adapt build-packages.sh to use Meson, it makes no sense to me having a build system configuration and not using it at all for build/package releases.

[1]: https://github.com/lite-xl/lite-xl/issues/307#issuecomment-867951940

redtide commented:
@franko am I doing something wrong here? <https://github.com/redtide/lite-xl/runs/3086628912>
[here](https://github.com/redtide/lite-xl/blob/ci-reorg-master/.github/workflows/build.yml#L71-L78) the failing part.

francesco-st commented:
> @franko am I doing something wrong here? https://github.com/redtide/lite-xl/runs/3086628912
> [here](https://github.com/redtide/lite-xl/blob/ci-reorg-master/.github/workflows/build.yml#L71-L78) the failing part.

I think I see the problem. "lhelper create lite-xl" creates a new environments and starts a new shell inside the new environment. The following commands "lhelper install ..." should be executed inside the sub-shell to work correctly. Now I think that the CI environment, when executing "lhelper install ..." has already closed the sub-shell where the environment was activated.

You should figure out how to continue and give the install commands from within the subshell. The sub-shell is started either with the "create" command for a new environment or using the "activate" command for an environment that already exists.

If needed I may implement some non-interactive option to provide the needed behavior.

redtide commented:
I suspected the same, but only after your answer I realized that I had to try to make it work from a shell script that might be executed in the same shell process. Unfortunately it doesn't work either, same as calling the commands from the CI YAML script.

redtide commented:
OK, I give up for a while waiting for some help later. Looking at the issue in general it's very unfortunate to have those bugs in sdl (ARM and WaitEvent). But in the end are all these static linking necessary? I mean why also PCRE2 and Freetype?

franko commented:
> I suspected the same, but only after your answer I realized that I had to try to make it work from a shell script that might be executed in the same shell process. Unfortunately it doesn't work either, same as calling the commands from the CI YAML script.

Ok, we may try the following approach. We create a script that we execute with bash:

```sh
cat << EOF > build-script.sh
# source the environment "lite-xl"
source "$LHELPER_PREFIX/environments/lite-xl"

# install the libraries in the environment
lhelper install freetype2
lhelper install sdl2 2.0.14-wait-event-timeout-1
lhelper install pcre2

# enter into lite-xl source directory and build the package
cd lite-xl
bash build-packages.sh x86-64
# in alternative we may use
# meson setup --buildtype=release .build && meson -C .build
EOF

# execute the script
bash build-script.sh
```

The only thing missing is that you have to figure out LHELPER_PREFIX which is the directory where lhelper is installed. For example is lhelper is in /usr/bin than LHELPER_PREFIX will be /usr.

With the approach above we are using two tricks:

1. instead of using "lhelper activate" or "lhelper create" we just "source" the environment into the current shell. This way we avoid entering into a sub-shell.
2. we run the sequence of commands including "source", the lhelper install commands and the build of lite-xl from a script file to ensure that they are all executed within the same shell session. In this way we are sure that the variables of the lhelper environments are configured.

For the moment this is the most reasonable solution that I see.

franko commented:
> OK, I give up for a while waiting for some help later. Looking at the issue in general it's very unfortunate to have those bugs in sdl (ARM and WaitEvent). But in the end are all these static linking necessary? I mean why also PCRE2 and Freetype?

Hey, there is no reason to give up now but I can understand your frustration. What we are doing is not easy so it is okay if it takes time and we need to clear some issues.

Static linking is not necessary but this is the practice I established for the releases and I don't see any strong reason to step back. The difficulties we are having with github actions are just a small technical problem we can solve.

Creating also a macos M1 release is also far from trivial so it is better to stay cool and take all the time need. Other projects have teams of full time engineers to work on these kind of tasks.

redtide commented:
After some attempt to find the path of lhelper environment I got <https://github.com/redtide/lite-xl/runs/3092467976?check_suite_focus=true#step:7:13>

I believe the traditional build is the most reproducible way in case of errors.
I still believe PCRE2 and Freetype could be used as external dependencies and link statically only what is needed:
- SDL2 until the patch is available, basically some years thanks to Debian :P 
- Lua: aren't recent version other than 5.2 suitable?

libagg is an old version with a more permissive license and unmaintained and reproc is not widely available as library so it makes sense to link them statically.

franko commented:
> After some attempt to find the path of lhelper environment I got https://github.com/redtide/lite-xl/runs/3092467976?check_suite_focus=true#step:7:13

That error may be something trivial but we need to inspect the logs files in /usr/local/.ghcup/var/lhenv/lite-xl/logs, freetype2-stdout.txt freetype2-stderr.txt etc. Probably there is only a single simple reason that explain why they are all failing.

redtide commented:
I wonder how to access those logs before the job ends and everything is deleted.

franko commented:
> I wonder how to access those logs before the job ends and everything is deleted.

This is why working with the github actions is a real pain.

redtide commented:
So I replaced the build [tarball][1] with a directory containing the errors and force the job to finish and upload even if failed.
There are 2 errors, 1 common for all builds:
`/usr/local/.ghcup/share/lhelper/recipes/pcre2_10.36: line 14: enter_archive: command not found` (the common one)
and `fatal: repository 'release-' does not exist` in `sdl2-stderr.log`.

Although LHelper makes things easier I think in this case it's adding another difficulty level to the CI build.
My suggestion is:
- Add a SDL2 subproject to Meson to facilitate the build (to remove later on when major distro have the new patched version)
- Dynamic link against `PCRE2` and `Freetype2`, and static link against `SDL2`, `libagg` and `reproc`. Not sure about `Lua`. This on Linux
- Same as above for MSYS2, including the mingw dlls + runtimes in the package
- Static link them all on macOS to permit the universal build

[1]: https://github.com/redtide/lite-xl/suites/3263251672/artifacts/75944131

franko commented:
Please let me solve this problem with lhelper. I will take the time to fix this, please be patient a little bit.

No reason to give up now.

redtide commented:
I'm not giving up, I just have no motivation now that the v2.0 will be released and I will have to wait another release to see some results, other than using a procedure that made things more complicated instead helping to finish a work I started weeks ago.

francesco-st commented:
I apologize for these inconvenient with lhelper but my intention was to follow the best practices and stay aligned with the release I was doing in the past. I think we can still be in time for the 2.0 release.

In any case it is clear to me that we are debugging lhelper to run in a CI environment. While this can be annoying for you I think it is an useful work that will pay in the future.

So, for the error we have in the logs, `/usr/local/.ghcup/share/lhelper/recipes/pcre2_10.36: line 14: enter_archive: command not found`, I don't quite understand how this can happen but I have two suggestions:

- check the bash version used in the CI environment, it could be a very old version and this may explain the problem
- use lhelper from the [direct-recipe-source branch](https://github.com/franko/lhelper/tree/direct-recipe-source). I simplified the script to avoid the offending "export -f" command. I made some tests and the modifications works finely so you may try this new branch in the CI.

Please note that in any case it is a good idea to check the bash version. Lhelper is heavily based on bash and we need a decently recent version.

redtide commented:
Sorry but at this point I'm against the use of a custom tool instead using the usual build system procedure. I faced already in the past in other contexts/projects problems like this when using an own tooling which other people (in the org or external contributors) have to learn and deal with, here it seems to try to solve a problem using what it appears as premature optimization.
This makes also hard to package on distributions/systems and my primary goal was to have an infrastructure to support the packaging on Archlinux.
Apart the SDL2 bug I still don't understand why not using dynamic linking for freetype and pcre2, which could make things much easier, optimization is not helping anyone here.
So for this reason I stop here, feel free to find a solution and integrate lhelper, ATM the most important thing is to have the CI working for all commits and PR, which is a great thing to have as it is available for free.

francesco-st commented:
You may continue in this case without using lhelper using standard tools. Please go ahead using dynamically linked libraries for SDL2, freetype2 and pcre2.

We may ship all these libraries as DLL in the windows' packages. Many people will not notice the CPU usage when the application is idle or reduced to icon.

For linux things are more complicated because each distribution has its own versions of each of the libraries so an executable that works for a distribution may not work for another one. May be with some care this can be made so that it works on most distributions.

On MacOS a special effort would be needed to package the additional dynamic libraries, for the moment I side-stepped the problem because I had no dependency on third party libraries.

To add SDL2 as a subproject is unfortunately not an option I can consider so if you go that way SDL2 will be a dynamic library that should be provided by each host system. SDL2 is too big and too complex to be added as a subproject.

redtide commented:
> You may continue in this case without using lhelper using standard tools. Please go ahead using dynamically linked libraries for SDL2, freetype2 and pcre2.

Dynamic linking may also speedup the build a bit, at least an installation should be quicker than a build!
But while I think this is more convenient and easier way for commits and PRs, I guess this might be not be ideal for you for releases, so I thought that we could either disable the release deployment or use different CI jobs by using lhelper if you want to work on them.

> We may ship all these libraries as DLL in the windows' packages. Many people will not notice the CPU usage when the application is idle or reduced to icon.

I wonder how much will take for SDL2 to release with your patch, I guess it should take just few days for having it available as package once released. Otherwise a possibility would be to make a patched PKGBUILD including your patch, build the package and add it in a custom pacman repository to be used in the CI until the new version comes out.

> For linux things are more complicated because each distribution has its own versions of each of the libraries so an executable that works for a distribution may not work for another one. May be with some care this can be made so that it works on most distributions.

I think we should keep track by documenting the possible issues on the website in some "known issues" page, I guess a problem could be on less recent Debian releases which uses old libraries, but having the same structure we can at least see when new problems arrives in the most updated distro, like on Arch being a rolling release, same for MSYS2.

> On MacOS a special effort would be needed to package the additional dynamic libraries, for the moment I side-stepped the problem because I had no dependency on third party libraries.

This should be solved using [dylibbundler](https://github.com/auriamg/macdylibbundler).
BTW, I forgot your POV about build all deps as static using AppVeyor for the universal builds, I don't think we can have both x86_64 and ARM64 libraries available to embed in the bundle, at least using brew (nor using the above tool).
Last thing: can we add your notarization data from the current dmg to the CI one?

francesco-st commented:
I want to be clear about what I have in mind:

- I still do not want to release lite-xl packages with bundled dynamic libraries
- I do not want to ship an AppImage
- I want to have a portable application zip package beside the "Setup" based one, if we get one.
- I don't see the problem with the packages I was providing in the previous release

I also would like to disable automatic CI action to build the package at each commit or PR.

I am glad you want to use and tinker with CI to automate builds and we may end up creating automatic packages for some scenarios but please do not try to impose your point of view about the releases.

I do not want to impose you to use lhelper but the points I listed above still holds.

francesco-st commented:
I was going to work on the github action to build lite-xl using lhelper but your version using lhelper it is no longer available in this branch because you force-pushed your changes.

Can you please provide me the last version you made using lhelper ?

For your information I pushed into lhelper master the fix about the missing function error. In addition I added a new lhelper's command, "env-source" so that you can "source" an environment without having to guess where lhelper is installed:

```sh
source "$(lhelper env-source <env-name>)"
```

where "<env-name>" should be "lite-xl" or whatever name you choose for the environment.

In this way we should simplify the github action script and just use lhelper from the master branch.

redtide commented:
> I also would like to disable automatic CI action to build the package at each commit or PR.

If I knew this since the beginning I'll never waste my time (weeks of works, not hours) for this.

> I am glad you want to use and tinker with CI to automate builds and we may end up creating automatic packages for some scenarios but please do not try to impose your point of view about the releases.

I'm not doing this if you read the other 2 replies above.

> I do not want to impose you to use lhelper but the points I listed above still holds.

It seems you are, because if no dynamic libraries can be used (and you told me to do it when I said to stop) and not even using subprojects (e.g. sdl2) I don't see any other option.

> Can you please provide me the last version you made using lhelper ?

I don't think to have it but if you click above on the previous commit (85a2579) it's still there, just append `.diff` to the url and you can apply it with `git apply`.

francesco-st commented:
> It's not mandatory that those packages have to replace the releases; I said above you can disable the release deployment until the packages doesn't satisfy/matches the current and/or desired method, but the packages/artifacts produced can still be available.

I think that automatic CI build are now done in the master branch maybe because of the previous PR we merged. I would just prefer that for non CI-related development work (push and PR in the master branch) the CI is not triggered.

> Apart the fact that there is nothing wrong to use dynamic linking, it is a simple and quick method to build. If lhelper is quicker than that and can be used directly also for releases I have nothing against it.

I agree, using dynamic libraries for quickly and easily build in a CI job is perfectly fine. My remark was about the released packages.

> > I do not want to impose you to use lhelper but the points I listed above still holds.
> 
> It seems you are, because if no dynamic libraries can be used (and you told me to do it when I said to stop) and not even using subprojects (e.g. sdl2) I don't see any other option.

Sorry I meant:

1. you can proceed using dynamic libraries for CI build automation but, at the end, I still don't want them for the release
2. for the release I don't force you to use lhelper but for the final package I prefer to not have additional shared libraries (so linking with static version of the library). Using sdl2 as a subproject is not an option we can consider, the library is too big and doesn't make sense.

Static library were not invented or discovered by me, they exists and can be used since compilers exist at least and many distributions support them out-of-the-box if you choose the right options. They often provides the static library besides the shared one in the dev package and pkg-config has options to select the static library. In term of libraries build, all of them support shared and static builds as an option.

Please don't make shortcut, using static libraries to link an executable is a standard well know practice and lhelper is not needed for that purpose.

> > Can you please provide me the last version you made using lhelper ?
> 
> I don't think to have it but if you click above on the previous commit ([85a2579](https://github.com/lite-xl/lite-xl/commit/85a257901264c4e200566334709fc8c29eb75d20)) it's still there, just append `.diff` to the url and you can apply it with `git apply`.

Thank you, I'll try that. It would be a pity if we lost this version of the CI build using lhelper because I wanted to use it to fix the problem you encountered and proceed along the direction we set.

francesco-st commented:
Just an adjustment about my previous comment about shipping an AppImage and related to https://github.com/lite-xl/lite-xl/discussions/342.

I said I don't want to ship an AppImage but I was too hasty to write this. In reality I think shipping an AppImage is a good thing but it should be built along the rules I set for the release packages so it should not use bundled dynamic libraries.

In other terms an AppImage should be built like an ordinary release package on linux with a final step that bundles everything in a single AppImage file.

redtide commented:
> I think that automatic CI build are now done in the master branch maybe because of the previous PR we merged. I would just prefer that for non CI-related development work (push and PR in the master branch) the CI is not triggered.

This is not clear, so you decided to not use CI that way? I don't understand in which way you are going to use it, I thought it was manual previously because incomplete, manual CI is completely useless and I don't see what's wrong if it is automatic, which to me is the main purpose of it.
And yes, it was part of #305, probably by mistake because I started to work on CI at that time.

> Static library were not invented or discovered by me, they exists and can be used since compilers exist at least and many distributions support them out-of-the-box if you choose the right options. They often provides the static library besides the shared one in the dev package and pkg-config has options to select the static library. In term of libraries build, all of them support shared and static builds as an option.

I didn't say anything against static libraries, I just don't agree it's necessary for wide used libraries like SDL, PCRE and Freetype in this context.
That might be Debian, not the case for Archlinux, OTOH it could be hard to maintain a system if anyone starts to static link anything like that with many different builds of the same libraries.
Lib agg and reproc ofc have sense to be static instead.

> Please don't make shortcut, using static libraries to link an executable is a standard well know practice and lhelper is not needed for that purpose.

I don't get what do you mean with shortcut?

> Thank you, I'll try that. It would be a pity if we lost this version of the CI build using lhelper because I wanted to use it to fix the problem you encountered and proceed along the direction we set.

I rewrote without it because I wanted to have a working result since I'm quite tired of this stuff.
If you have a lhelper procedure that works on CI please post it so I can force push again to include it.

redtide commented:
> For your information I pushed into lhelper master the fix about the missing function error. In addition I added a new lhelper's command, "env-source" so that you can "source" an environment without having to guess where lhelper is installed:
> 
> ```shell
> source "$(lhelper env-source <env-name>)"
> ```

<https://github.com/redtide/lite-xl/runs/3154820246#step:6:22>, script is [here](https://github.com/redtide/lite-xl/blob/ci-reorg-master-private/scripts/run-lhelper.sh)

franko commented:
> > For your information I pushed into lhelper master the fix about the missing function error. In addition I added a new lhelper's command, "env-source" so that you can "source" an environment without having to guess where lhelper is installed:
> > ```shell
> > source "$(lhelper env-source <env-name>)"
> > ```
> 
> https://github.com/redtide/lite-xl/runs/3154820246#step:6:22, script is [here](https://github.com/redtide/lite-xl/blob/ci-reorg-master-private/scripts/run-lhelper.sh)

Ciao Andrea!

sorry I messed up with the install-github script which was outdated and was installing an old version of lhelper. Now I updated the script to always install the latest version.

With the lhelper install problem solved I tested your script locally on my PC and it works.

I only had to add the "-n" flag when using the "lhelper create lite-xl" command. The -n flag tells to lhelper to not start a subshell into the newly created environment. We don't want this because we are running non interactively.

Here the script I used, I added also a command to build lite-xl but I guess you will plug your own version.

```
#!/bin/bash

git clone https://github.com/franko/lhelper.git
pushd lhelper; bash install-github; popd

lhelper create lite-xl -n

# Not using `lhelper activate lite-xl`

#source "/usr/local/.ghcup/var/lib/lhelper/environments/lite-xl"
source "$(lhelper env-source lite-xl)"

lhelper install freetype2
lhelper install sdl2 2.0.14-wait-event-timeout-1
lhelper install pcre2

git clone https://github.com/lite-xl/lite-xl.git lite-xl-gh-build

cd lite-xl-gh-build
bash build-packages.sh x86-64

#meson setup --buildtype=release --prefix /usr build
#meson compile -C build
```

redtide commented:
That worked, thanks! I'll adapt all builds that way then.
Meanwhile I'm having some strange issues with MSYS2, so I'm tempted (actually I already started) to use a Docker container and build on Windows using an Archlinux image with MinGW instead (which it's very similar as using MSYS2), the only problem this will require to create a repository with PKGBUILDs and related binary packages to use with it, so self prebuilt sdl2 and friends.

Can I include your notarization data file (and related references in the plist file) in the current dmg image, to match the one you are packaging? If yes may be the case to include them in Lite XL repository macos scripts directory?

franko commented:
> That worked, thanks! I'll adapt all builds that way then.

It's nice that we got this working at last. One advantage with lhelper the recipe is mostly the same across Windows (using MSYS2), macOS and linux.

You just need to build the environment a little bit differently on macOS because SDL2 needs to be compiled with some opengl specific options. I can give you the details if you want. You need also to set CC=clang CXX=clang++ when creating the environment like:

```sh
CC=clang CXX=clang++ lhelper create lite-xl -n
```

> Meanwhile I'm having some strange issues with MSYS2, so I'm tempted (actually I already started) to use a Docker container and build on Windows using an Archlinux image with MinGW instead (which it's very similar as using MSYS2), the only problem this will require to create a repository with PKGBUILDs and related binary packages to use with it, so self prebuilt sdl2 and friends.

I use MSYS2 to build lite-xl on Windows and I have no problem. Actually I use tdm-gcc with msys2 to produce the release builds but this is a detail and the msys2 provided gcc works as well. You can even use clang with msys2.

If you share with me the errors you got with MSYS2 maybe I can help you. I think it is worthwhile to have a working github action job for MSYS2.

> Can I include your notarization data file (and related references in the plist file) in the current dmg image, to match the one you are packaging? If yes may be the case to include them in Lite XL repository macos scripts directory?

Well, the problem is that to notarize you need your secret key and the password to unlock it and I think I cannot share them. I propose that for the moment we exercise the build pipeline to create an package which is not notarized.

> Would be also the case to make the Linux build portable?

I think I don't understand what you mean but I thought it would be nice to add a section in build-helper.sh to create an AppImage using the build we already make with static libraries etc using lhelper.

As for the portability on linux, sometimes it is enough to build the binaries in an old ubuntu release to link with an older version of libc. Otherwise you can static-link libc for example using musl-lic but this is very radical!
 
> I'm getting the same old errors in the [macOS build](https://github.com/redtide/lite-xl/runs/3160091940#step:6:20).

Strange! I will try to have a look but on macos you need to install a recent bash version using homebrew and ensure that we use the new version.

redtide commented:
> I can give you the details if you want.

ofc I want them, otherwise anything explode here :-P I'll try your suggestions for the macOS build.
BTW I also have no problems with MSYS2 in my VM, but unfortunately I have no idea about the crap happening [here][1], I also joined the meson discord channel and no help even there, I'll retry later using lhelper.

[1]: https://github.com/redtide/lite-xl/runs/3150856309#step:8:36

> I propose that for the moment we exercise the build pipeline to create an package which is not notarized.

Fine to me, BTW there is a project I use as reference that does codesigning using Actions, if you trust it. We can work about this in a later moment anyway if the case.

> I think I don't understand what you mean but I thought it would be nice to add a section in build-helper.sh to create an AppImage using the build we already make with static libraries etc using lhelper.

I meant about the current Linux binary tarball that seems unusable if not installed, so we could make a portable build with my current patched meson.build instead, so no bin/share directories, with the executable in the root.
It would be great if you can adapt lhelper to add AppImage compatibility so that same build will be used for both portable and AppImage packages.

redtide commented:
<https://github.com/redtide/lite-xl/runs/3161512918#step:6:10> bash works but it seems macOS doesn't come with md5sum tool installed, it has md5 instead.

OT: sorry to chat here but it's a quick question: someone asked for a link to the website from the info of this project. Can you please add it with some tags and set the organization icon when you have time?

redtide commented:
OK, I guess I can make it work by just adapting the appimage.sh script.

About MSYS2 I get an error while building because SDL2.h include not found, which IDK why. <https://github.com/redtide/lite-xl/runs/3163314826#step:6:324>

franko commented:
> https://github.com/redtide/lite-xl/runs/3161512918#step:6:10 bash works but it seems macOS doesn't come with md5sum tool installed, it has md5 instead.

Well, the precondition to work on macOS with lhelper, meson etc is to install up to date unix tools using homebrew. Unfortunately I didn't write down the packages I had to install but md5sum was one of them. I think including a recent bash there isn't much else that is needed.

> MSYS2 gives log errors https://github.com/redtide/lite-xl/runs/3162387844

You may try to use the "build-release.sh" script. I don't see why you don't start from that since I already made some work to make it work correctly. The error could be a problem with paths. Msys is tricky because sometimes wants /c/somedir and sometimes C:/somedir. I baked the solution for many problems of this kind inside lhelper.

> OT: sorry to chat here but it's a quick question: someone asked for a link to the website from the info of this project. Can you please add it with some tags and set the organization icon when you have time?

No problem to do this in principle but unfortunately I have very little time in this moment.

redtide commented:
> I think including a recent bash there isn't much else that is needed.

As I said previously the new bash was installed and working by testing with `bash --version` before and after the installation, but I still get errors.

> You may try to use the "build-release.sh" script. I don't see why you don't start from that since I already made some work to make it work correctly.

- it misses some things I added with other contributors suggestions
- it is customized in a way it doesn't use Meson in some cases
- I don't entirely understand it so I have no control on what it does

redtide commented:
So in the end what I got until now using LHelper are:
- a working Linux (excluding AppImage, need to adapt the script to LHelper) and macOS builds
- a broken compilation on MSYS2, with or without using `build-packages.sh`
- each build can take up to 6 minutes vs. even less than a minute using installed dlls

so what I see here is to have:
- LHelper builds used only for releases if/when you want them to be built automatically on CI
- using traditional dlls for CI commits and pull requests if you don't decide to disable them

franko commented:
> So in the end what I got until now using LHelper are:
> 
>     * a working Linux (excluding AppImage, need to adapt the script to LHelper) and macOS builds
> 
>     * a broken compilation on MSYS2, with or without using `build-packages.sh`

Msys2 was working before without lhelper ? That's funny because I first created lhelper to help me compile thing on windows with msys. Before lhelper it was a pain for me to compile stuff on Windows.

>     * each build can take up to 6 minutes vs. even less than a minute using installed dlls

Right! This happens because the CI thing send the lhelper environment into the trashbin the moment you finished the job. In the normal utilization you keep the environment so following builds don't have to rebuild the libraries.

In practice this means that lhelper is an option for creating high-quality release packages but for CI builds it is better to use the system provided libraries, shared or not.

> so what I see here is to have:
> 
>     * LHelper builds used only for releases if/when you want them to be built automatically on CI
> 
>     * using traditional dlls for CI commits and pull requests if you don't decide to disable them

So we agree about that except that I maintain that we don't need CI builds for lite-xl at each commit and for each PR.

Most of the changes we made are about the Lua code, we don't touch that often the C/C++ part of the codebase. In addition we don't have any unit or integration test so the CI add almost no added value for our project, it is confusing (the CI build is failing but is it related to my PR ?) and wastes CPU resources for nothing.

The work you are doing for CI builds is useful but we don't need to make CI builds all the time with no reason.

Maybe we can setup something intermediate as nightly builds or something like that. Since the development pace of lite-xl is not crazy high we may do weekly builds or something like that. For these kind of development builds it is fine to me to not use lhelper and use a standard approach with system provided libraries. It could be nice to provide these development builds in our website for people who wants to test some recent features.

For the lhelper based build I see this as an valuable asset for our project to produce high-quality builds without the need of having access to a specific machine.

redtide commented:
> Msys2 was working before without lhelper ? That's funny because I first created lhelper to help me compile thing on windows with msys. Before lhelper it was a pain for me to compile stuff on Windows.

yes, you see it here in the passed checks in the current state of the PR, which is not using LHelper at all yet. At some point I had both 32 and 64 bit working, but ATM only 64 bits builds correctly, using LHelper I get a broken build because the SDL.h not found while building with LHelper (see the logs, in my [private branch build][1] you have also [error logs uploaded][2]). I'm sure it's something that can be fixed but I have no idea how ATM.
I'm not sure which shell is used here, I never use msys shell but only mingw ones on my system.

The fact that most of the code is made with Lua is not relevant if not the fact that makes the builds made quicker.
It's already planned on Actions to build on each commit/push, otherwise it would be forbidden, and in another complex c++ project I taken for reference since the beginning builds works like this (for free) since months.
Tests can be added in a later time, so if that makes sense for you to use CI why not now?
I don't consider builds using dlls as kind of nightly releases, but temporary build artifacts which are useful for users to test at any time.
Anyway your project, your rules, this is what I wanted and could do.

[1]: https://github.com/redtide/lite-xl/actions/runs/1068498861
[2]: https://github.com/redtide/lite-xl/suites/3333828554/artifacts/78184782

redtide commented:
I made some changes to adapt the Linux AppImage script to lhelper, but I'm not sure how the portable option works, this to understand how to build once and use the same build for both AppImage and the portable archive.
When I asked if we could release a portable build I meant about using a different directory structure from the posix one used in the current releases but I didn't know that even as is now, with the executable in `bin` instead in the root of the archive directory, it works as local portable. Someone else kindly confirmed this. So now I'm not sure which directory structure should be used if both are portable.
What is the reason for the portable option then?

franko commented:
> When I asked if we could release a portable build I meant about using a different directory structure from the posix one used in the current releases but I didn't know that even as is now, with the executable in `bin` instead in the root of the archive directory, it works as local portable. Someone else kindly confirmed this. So now I'm not sure which directory structure should be used if both are portable.
> What is the reason for the portable option then?

The compilation of the executable and the executable itself are not affected by the "portable" option. The executable automatically detect if it is used in a unix-style installation but looking if it is inside a "bin" directory. If yes it looks for the Lua modules in ../share/lite-xl. If it is not in a "bin" directory it assumes it is installed in a portable way and just look for a "data" directory next to the executable itself.

So the "portable" option is a remnant of an old configuration and is currently used only to determine where we should install the files when we run "meson install". This means that we can build the executable only once and create the portable and unix-like packages using the same executable. This also means that, when using the unix-like install you can install it in any prefix directory of your choice, for example /usr, /usr/local, $HOME/.local or any other directory you may choose.

I thought that it was stupid to hard-code the installation path into the executable. It was also easy to avoid hard-coding the "portable" option as well.

redtide commented:
So on Linux the portable option is irrelevant and the unix style as is now is already portable? So there is no reason to use the file in the root as someone expect from a portable build. So I'll get rid of it also on MSYS2 then.

franko commented:
> So on Linux the portable option is irrelevant and the unix style as is now is already portable? So there is no reason to use the file in the root as someone expect from a portable build. So I'll get rid of it also on MSYS2 then.

Not sure I understand your comment but the idea is that some people may prefer a portable install (all in one directory) and other an unix-like one. The executable is the same, what changes is just how we package the files.

> Another question: can we include also the desktop and icon on Unix in the Meson install? It seems it is not there like for other platforms.

Well, of course we can! I was not very attentive to the meson install but you are right about fixing this.

redtide commented:
> Not sure I understand your comment but the idea is that some people may prefer a portable install (all in one directory) and other an unix-like one. The executable is the same, what changes is just how we package the files.

In brief I want to know how to package, using the portable structure or the current unix like, otherwise having all of them with 3 packages, counting the appimage too.

redtide commented:
FWIW I solved the issue with MSYS2 by passing the SDL2 include directory via CFLAGS and library via LDFLAGS, IDK why it happens only on MSYS2, the `LHELPER_ENV_PREFIX` looks like as the same on Linux.
I will get rid of the win32 build for now.
I've added quite few scripts and changed some files, maybe will be the case to split this PR with other commits on different PR, let me know.

franko commented:
> In brief I want to know how to package, using the portable structure or the current unix like, otherwise having all of them with 3 packages, counting the appimage too.

Well, my idea for linux was to provide an unix-like package and the appimage. It looks reasonable to me as it cover the major use case.

franko commented:
> FWIW I solved the issue with MSYS2 by passing the SDL2 include directory via CFLAGS and library via LDFLAGS, IDK why it happens only on MSYS2, the `LHELPER_ENV_PREFIX` looks like as the same on Linux.
> I will get rid of the mingw 32 bit build for now.
> I've added quite few scripts and changed some files, maybe will be the case to split this PR with other commits on different PR, let me know.
> I'm now adjusting the workflow file in a way that it will use lhelper on releases and install and use dll on normal commits as planned, without need to duplicate jobs. If you'll decide to disable something can be made in few lines without the need to delete my work.

From what you are saying looks like a bug in lhelper. If you share the details I may have a look, this is maybe something I can fix.

redtide commented:
> From what you are saying looks like a bug in lhelper. If you share the details I may have a look, this is maybe something I can fix.

I have no idea what to share other than the error logs I linked [above](https://github.com/lite-xl/lite-xl/pull/336#issuecomment-887019391) on the previous broken build.
Anyway it is working and I see no harm by passing meson those paths, so I will focus on finishing the PR.

franko commented:
> I have no idea what to share other than the error logs I linked [above](https://github.com/lite-xl/lite-xl/pull/336#issuecomment-887019391) on the previous broken build.
> Anyway it is working and I see no harm by passing meson those paths, so I will focus on finishing the PR.

Ok, looking at the logs I confirm this is a bug in lhelper but for the moment it is not clear to me what is going wrong.

To understand the problem I would need the file sdl2-config generated by the CI build in `D:\a\_temp\msys\msys64\usr\var\lhenv\lite-xl\bin/sdl2-config` if possible.

redtide commented:
> To understand the problem I would need the file sdl2-config generated by the CI build in `D:\a\_temp\msys\msys64\usr\var\lhenv\lite-xl\bin/sdl2-config` if possible.

I'll add it to the error logs and make a failing build to generate them.

Questions:
1. What do you think if I split this PR? It seems too much changes to me for 1 only.
2. Can we get rid of the unix like build and make only the installer and a portable build with the exe in the main directory? Despite MSYS2/MinGW is a unix based environment the OS is always Windows, I don't think much people care about a package with unix like directory tree. And if yes I would like to propose to remove the `portable` switch since the application is already portable whatever structure/OS is using, so it's just confusing, and set the portable tree only for Windows.

redtide commented:
OK, so I got what it is about the final version of this PR, but I need some help to try to fix the MSYS2 issues with MINGW32.
Does anyone know what is the cause of this error or have experienced something similar? It looks quite odd to me, like meson is not found when running meson? Both packages are installed:

```
meson setup --buildtype=release --prefix / --wrap-mode=forcefallback build
Traceback (most recent call last):
  File "C:\msys64\mingw32\bin\meson-script.py", line 33, in <module>
    sys.exit(load_entry_point('meson==0.58.1', 'console_scripts', 'meson')())
  File "C:\msys64\mingw32\bin\meson-script.py", line 22, in importlib_load_entry_point
    for entry_point in distribution(dist_name).entry_points
  File "C:/msys64/mingw32/lib/python3.8/importlib/metadata.py", line 503, in distribution
    return Distribution.from_name(distribution_name)
  File "C:/msys64/mingw32/lib/python3.8/importlib/metadata.py", line 177, in from_name
    raise PackageNotFoundError(name)
importlib.metadata.PackageNotFoundError: meson
```
```
$ pacman -Q ${MINGW_PACKAGE_PREFIX}-meson ${MINGW_PACKAGE_PREFIX}-ninja
mingw-w64-i686-meson 0.58.1-2
mingw-w64-i686-ninja 1.10.2-1
```

redtide commented:
@franko sorry again but I think I need to talk to you a bit in chat if you have time this weekend.

redtide commented:
Closing, replaced by #360, #361, #362, #372 and #373.
